### PR TITLE
HDDS-2300. Publish normalized Ratis metrics via the prometheus endpoint 

### PR DESCRIPTION
**Pull request by elek**
## What changes were proposed in this pull request?

Latest Ratis contains very good metrics about the status of the ratis ring.

After RATIS-702 it will be possible to adjust the reporter of the Dropwizard based Ratis metrics and export them directly to the /prom HTTP endpoint (used by ozone insight and ratis).

Unfortunately Dropwizard is very simple, there is no tag support. All of the instance specific strings are part of the metric name. For example:

```
"ratis_grpc.log_appender.72caaf3a-fb1c-4da4-9cc0-a2ce21bb8e67@group"
 + "-72caaf3a-fb1c-4da4-9cc0-a2ce21bb8e67"
 + ".grpc_log_appender_follower_75fa730a-59f0-4547"
 + "-bd68-216162c263eb_latency", 
```

In this patch I will use a simple method: during the export of the dropwizard metrics based on the well-known format of the ratis metrics, they are converted to proper prometheus metrics where the instance information is included as tags:

```
ratis_grpc.log_appender.grpc_log_appender_follower_latency{instance="72caaf3a-fb1c-4da4-9cc0-a2ce21bb8e67"}
 ```

With this approach we can:

  1. monitor easily all the Ratis pipelines with one simple query
  2. Use the metrics for ozone insight which will show health state of the Ratis pipeline

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-2300

## How was this patch tested?

 1. Start an ozoneperf docker-compose cluster.
 2. scale up to 3 datanodes
 3. do something (eg. freon or key upload)
 4. Check http://localhost:9090 prometheus endpoint
 5. Check the metrics with 'ratis_' prefix. There shouldn't be any unique id in the name of the metrics.

### CODE REVIEW

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/PrometheusServlet.java [hunk 1]
```diff
@@ -40,7 +44,12 @@ public PrometheusMetricsSink getPrometheusSink() {
   protected void doGet(HttpServletRequest req, HttpServletResponse resp)
       throws ServletException, IOException {
     DefaultMetricsSystem.instance().publishMetricsNow();
-    getPrometheusSink().writeMetrics(resp.getWriter());
-    resp.getWriter().flush();
+    PrintWriter writer = resp.getWriter();
+    getPrometheusSink().writeMetrics(writer);
+    resp.getWriter().write("\n\n#Dropwizard metrics\n\n");
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/PrometheusServlet.java]:
```suggestion
    writer.write("\n\n#Dropwizard metrics\n\n");
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/RatisNameRewriteSampleBuilder.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+              additionalLabelValues, value);
+    }
+  }
+
+  protected String normalizeRatisMetric(String dropwizardName,
+      List<String> names,
+      List<String> values) {
+
+    List<String> nameParts =
+        new ArrayList<String>(Arrays.asList(dropwizardName.split("\\.")));
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/RatisNameRewriteSampleBuilder.java]:
```suggestion
        new ArrayList<>(Arrays.asList(dropwizardName.split("\\.")));
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/RatisNameRewriteSampleBuilder.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    for (int i = 0; i < parts.length; i++) {
+      if (i != 2) {
+        if (i != 0) {
+          builder.append(".");
+        }
+        builder.append(parts[i]);
+      }
+    }
+    return builder.toString();
+  }
```
Comment (adoroszlai) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/RatisNameRewriteSampleBuilder.java]: Seems to be unused.

Comment (elek) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/RatisNameRewriteSampleBuilder.java]: thanks. deleted.

**Comment by elek** [comment]
Acceptance test failure (haproxy) is independent I created a separated issue to track it: HDDS-2606

**Comment by arp7** [comment]
+1

**Pull request merged**